Case claim request: Return 204 based on user's sync token

AddisonDunn opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

Invisible.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

Came out of the PR discussion [here](https://github.com/dimagi/formplayer/pull/1117). Currently, if a user has already claimed a case but tries to claim that case again, HQ will return a 204 code to tell formplayer/commcare that the user's device doesn't need to sync.

This PR changes that 204 code to be returned based on whether the case is already on the user's phone or not, instead of based on whether the user has previously claimed this case. This way, if a user happens to claim the case on device A and then do it again on device B without syncing, the sync will still (as expected) be triggered instead of HQ assuming that case is already on device B.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

SEARCH_CLAIM

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

Local testing w/ the web app. 204 is returned when a case has already been claimed, 201 when it is being newly claimed (as expected). I tried testing using the CLI too to test the potential multiple-devices issue described above but the CLI's case search workflow doesn't work w/ a restore file.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

None

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

None

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AddisonDunn committed:
return 204 based on sync token

orangejenny reviewed (commented):
Code looks good. Any idea how complex it would be to add a test?

AddisonDunn commented:
@orangejenny I was skeptical about putting in a test because I'd need to replicate the restore file. But I think I might be able to by borrowing some stuff from the [OTA restore test](https://github.com/dimagi/commcare-hq/blob/202fb2431c289ef8921b58ea27f7e3785b6cf1a7/corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore_v3.py#L15-L41).

orangejenny commented:
cool, nice find

snopoke commented on corehq/apps/ota/views.py:
```diff
@@ -157,8 +157,14 @@ def claim(request, domain):
-        return HttpResponse(status=204)
+    if not case_ids_to_claim:
+        from casexml.apps.phone.models import get_properly_wrapped_sync_log
+        synclog = get_properly_wrapped_sync_log(request.last_sync_token)
```
This needs to handle a `MissingSyncLog` error.

snopoke commented on corehq/apps/ota/views.py:
```diff
@@ -157,8 +157,14 @@ def claim(request, domain):
+        from casexml.apps.phone.models import get_properly_wrapped_sync_log
+        synclog = get_properly_wrapped_sync_log(request.last_sync_token)
+        missing_case_ids_on_phone = set(case_ids) - synclog.case_ids_on_phone
+        cases_already_on_phone = not(bool(missing_case_ids_on_phone))
```
nit: extracting this to a function would make it much cleaner
```python
if not case_ids_to_claim and phone_holds_all_cases(request.last_sync_token, case_ids):
    return HttpResponse(status=204)
```

snopoke commented on corehq/apps/ota/views.py:
```diff
@@ -157,8 +157,14 @@ def claim(request, domain):
-    if set(case_ids) == case_ids_already_claimed:
-        return HttpResponse(status=204)
+    if not case_ids_to_claim:
+        from casexml.apps.phone.models import get_properly_wrapped_sync_log
```
Is the inline import necessary?

AddisonDunn replied:
When should we import inline vs. at the top?

snopoke replied:
Imports should always be at the top unless there's a specific reason not to

AddisonDunn committed:
refactoring and catch MissingSincLog exception

AddisonDunn committed:
fix tests using single SyncLogSQL object

AddisonDunn commented:
See my new commits! Added the changes you suggested @snopoke. And fixed some breaking tests using a SyncLogSQL object. I think those existing tests should be sufficient @orangejenny.

snopoke commented on corehq/apps/ota/tests/test_search_claim_endpoints.py:
```diff
@@ -116,7 +128,8 @@ def test_duplicate_client_claim(self):
         # Dup claim
-        response = self.client.post(self.url, {'case_id': self.case_id})
+        response = self.client.post(self.url, {'case_id': self.case_id},
+            HTTP_X_COMMCAREHQ_LASTSYNCTOKEN=self.synclog.synclog_id)
```
can you add tests for duplicate claim with a synclog ID that doesn't exist and with a synclog that doesn't contain the case

snopoke reviewed (commented):
Almost there

kaapstorm commented on corehq/apps/ota/tests/test_search_claim_endpoints.py:
```diff
@@ -116,7 +128,8 @@ def test_duplicate_client_claim(self):
+            HTTP_X_COMMCAREHQ_LASTSYNCTOKEN=self.synclog.synclog_id)
         self.assertEqual(response.status_code, 204)
 
     def test_duplicate_user_claim(self):
```
I'm guessing the second test was supposed to test the "if a user happens to claim the case on device A and then do it again on device B" scenario.

snopoke replied:
OK, that scenario (kind of) makes sense but the test isn't actually doing anything to allow the view to know that it's a different device (as opposed to another request from the same device).

I think it's safe to delete this test then. The test I've recommended in https://github.com/dimagi/commcare-hq/pull/31458#discussion_r850142010 will cover the cases for requests coming from different devices.

AddisonDunn replied:
Done!

AddisonDunn committed:
various small changes

AddisonDunn committed:
modify test and add test to use synclog IDs

AddisonDunn commented:
@snopoke all updates made, see the latest!

AddisonDunn commented on corehq/apps/ota/tests/test_search_claim_endpoints.py:
```diff
@@ -86,11 +90,19 @@ def setUp(self):
         ensure_index_deleted(CASE_SEARCH_INDEX)
         self.user.delete(self.domain.name, deleted_by=None)
         self.domain.delete()
+        SyncLogSQL.objects.all().delete()
```
Ah, thanks!

snopoke commented on corehq/apps/ota/tests/test_search_claim_endpoints.py:
```diff
@@ -1,14 +1,16 @@
 
 from django.test import Client, TestCase
 from django.urls import reverse
+from custom.covid.tests.test_prime_restore import make_synclog
```
This utils method doesn't really do very much and it's in a custom module so not great to import into core code.

I think it would be fine to make a local function that applies any defaults for these tests.

If the existing function contained logic or something more complex then I'd suggest moving it to a more central location (like somewhere alongside the Synclog model itself).

AddisonDunn replied:
That makes sense. I added a custom make_synclog() function to CaseClaimEndpointTests in a new commit, can you give it a look?

snopoke reviewed (approved):
Looks good. Left one comment but not blocking.

AddisonDunn committed:
turn make_synclog() into a classmethod
